plugins/spectate/spectate.cpp

```diff
@@ -0,0 +1,382 @@
+            } else if (parameters[1] == "focus-jobs") {
+                focus_jobs_enabled = state;
+            } else if (parameters[1] == "tick-threshold" && set && parameters.size() == 3) {
+                try {
+                    tick_threshold = std::stol(parameters[2]);
+                } catch (const std::exception &e) {
+                    out.printerr("%s\n", e.what());
+                }
+            } else {
+                return DFHack::CR_WRONG_USAGE;
+            }
+        }
+    } else {
+        out.print("Spectate is %s\n", enabled ? "ENABLED." : "DISABLED.");
+        out.print("tick-threshold: %" PRIu64 "\n", tick_threshold);
+        out.print("focus-jobs: %s\n", focus_jobs_enabled ? "on." : "off.");
+        out.print("auto-unpause: %s\n", unpause_enabled ? "on." : "off.");
+        out.print("auto-disengage: %s\n", disengage_enabled ? "on." : "off.");
+    }
+    saveConfig();
+    return DFHack::CR_OK;
+}
+
+// every tick check whether to decide to follow a dwarf
+void onTick(color_ostream& out, void* ptr) {
+    if (!df::global::ui) return;
+    int32_t tick = df::global::world->frame_counter;
+    if (our_dorf) {
+        if (!Units::isAlive(our_dorf)) {
+            following_dwarf = false;
+            df::global::ui->follow_unit = -1;
+        }
+    }
+    if (!following_dwarf || (focus_jobs_enabled && !job_watched) || (tick - timestamp) > (int32_t) tick_threshold) {
+        std::vector<df::unit*> dwarves;
+        for (auto unit: df::global::world->units.active) {
+            if (!Units::isCitizen(unit)) {
+                continue;
+            }
+            dwarves.push_back(unit);
+        }
+        std::uniform_int_distribution<uint64_t> follow_any(0, dwarves.size() - 1);
+        if (df::global::ui) {
+            // if you're looking at a warning about a local address escaping, it means the unit* from dwarves (which aren't local)
+            our_dorf = dwarves[follow_any(RNG)];
+            df::global::ui->follow_unit = our_dorf->id;
+            job_watched = our_dorf->job.current_job;
+            following_dwarf = true;
+            if (!job_watched) {
+                timestamp = tick;
+            }
+        }
+    }
+}
+
+// every new worked job needs to be considered
+void onJobStart(color_ostream& out, void* job_ptr) {
+    // todo: detect mood jobs
+    int32_t tick = df::global::world->frame_counter;
+    auto job = (df::job*) job_ptr;
+    // don't forget about it
+    int zcount = ++freq[job->pos.z];
+    job_tracker.emplace(job->id);
+    // if we're not doing anything~ then let's pick something
+    if ((focus_jobs_enabled && !job_watched) || (tick - timestamp) > (int32_t) tick_threshold) {
+        following_dwarf = true;
+        // todo: allow the user to configure b, and also revise the math
+        const double b = base;
+        double p = b * ((double) zcount / job_tracker.size());
+        std::bernoulli_distribution follow_job(p);
```

Comment:
I like how you've put the random number generator engines to good use in this plugin : )

Reply:
Thanks. I did write the cxxrandom plugin